File: spec/e2e/foundation.spec.ts
import { expect, test } from '@playwright/test'

test('opens the sign-in page and performs an Inertia visit', async ({ page }) => {
  const errors: string[] = []
  page.on('pageerror', error => errors.push(error.message))
  page.on('console', message => { if (message.type() === 'error') errors.push(message.text()) })
  await page.goto('/')
  await expect(page).toHaveTitle('Entrar · Umanni')
  await expect(page.locator('html')).toHaveAttribute('lang', 'pt-BR')
  await expect(page.getByRole('heading', { name: 'Entre na sua conta', level: 1 })).toBeVisible()
  const visit = page.waitForResponse(response => response.request().headers()['x-inertia'] === 'true')
  await page.getByRole('link', { name: 'Cadastre-se' }).click()
  const response = await visit
  expect(response.status()).toBe(200)
  const body = await response.json()
  expect(body.component).toBe('Auth/SignUp')
  expect(body.props).toEqual({ auth: { user: null }, errors: {}, flash: {} })
  expect(errors).toEqual([])
})

test('reloads the document after a real asset version mismatch', async ({ page }) => {
  await page.goto('/')
  await page.route('**/*', async route => {
    const headers = route.request().headers()
    if (headers['x-inertia'] === 'true') headers['x-inertia-version'] = 'outdated'
    await route.continue({ headers })
  })
  const conflict = page.waitForResponse(response => response.status() === 409)
  const reload = page.waitForRequest(request => request.isNavigationRequest())
  await page.getByRole('link', { name: 'Cadastre-se' }).click()
  expect((await conflict).headers()['x-inertia-location']).toBeTruthy()
  await reload
  await expect(page.getByRole('heading', { name: 'Crie sua conta' })).toBeVisible()
})

File: spec/requests/foundation_spec.rb
require "rails_helper"

RSpec.describe "Foundation", type: :request do
  it "restricts executable assets to the application origin" do
    get "/"
    policy = response.headers.fetch("Content-Security-Policy", "")
    expect(policy).to include("default-src 'self'", "script-src 'self'", "object-src 'none'")
  end

  it "keeps the Rails session readable across consecutive browser visits" do
    get "/"
    expect(response).to have_http_status(:ok)

    get "/", headers: { "X-Inertia" => "true", "X-Inertia-Version" => ViteRuby.digest }
    expect(response).to have_http_status(:ok)
  end

  it "serves only the approved page props on an Inertia visit" do
    get "/", headers: { "X-Inertia" => "true", "X-Inertia-Version" => ViteRuby.digest }

    expect(response).to have_http_status(:ok)
    expect(response.headers["X-Inertia"]).to eq("true")
    expect(response.headers["Vary"]).to include("X-Inertia")
    expect(response.parsed_body).to include("component" => "Auth/SignIn", "url" => "/")
    expect(response.parsed_body.fetch("props")).to eq(
      "auth" => { "user" => nil }, "errors" => {}, "flash" => {}
    )
    expect(response.parsed_body.fetch("version")).to be_present
  end

  it "requires a full reload when the asset version differs" do
    get "/", headers: { "X-Inertia" => "true", "X-Inertia-Version" => "outdated" }

    expect(response).to have_http_status(:conflict)
    expect(response.headers["X-Inertia-Location"]).to eq("http://www.example.com/")
  end

  it "reports application boot at the health endpoint" do
    get "/up"
    expect(response).to have_http_status(:ok)
  end

  it "returns not found for unknown routes" do
    get "/missing-foundation-route"
    expect(response).to have_http_status(:not_found)
  end

  it "preserves CSRF protection and supplies a token without adding page props" do
    get "/"
    expect(ApplicationController.allow_forgery_protection).to be(true)
    expect(response.body).to include('name="csrf-token"')
    expect(response.cookies.fetch("XSRF-TOKEN")).to be_present
  end

  it "renders the initial Portuguese HTML document" do
    get "/"

    expect(response).to have_http_status(:ok)
    expect(response.body).to include('lang="pt-BR"', "Auth/SignIn", "data-page", "/vite-test/")
  end
end
